[REEF-1357] Allow different caching levels for caching in IInputParti…

afchung opened the pull request:
…tion

This addressed the issue by
  * Adding the interface IInputDataMover, which abstracts the movement of data between different memory layers.
  * Adding the DataCache class, which abstracts the caching of data at different storage layers.
  * Changing the IInputPartition interface to work with the new ISerializer.

JIRA:
  [REEF-1357](https://issues.apache.org/jira/browse/REEF-1357)

afchung commented:
@dkm2110 could you please take a look at this PR when you have time? Thanks!

dkm2110 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+        /// <summary>
+        /// The directory where cached data resides, if the data is cached on disk.
+        /// </summary>
+        public Optional<IDirectoryInfo> DiskDirectory
```
Question: why do these storage variables be public?

afchung replied:
They don't have to be, I'll hide them.

dkm2110 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+        /// Cache to memory with the injected <see cref="IInputDataMover{T}"/>.
+        /// </summary>
+        /// <returns>The cached level.</returns>
+        public CacheLevel CacheToMemory(bool shouldCleanHigherLevelCache)
```
Why should this be public?

afchung replied:
Sorry, I'll hide it.

dkm2110 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+                case CacheLevel.Disk:
+                    CheckDisk(true);
+                    materialized = Optional<IReadOnlyCollection<T>>.Of(
+                        new List<T>(_inputDataMover.DiskToMaterialized(_diskDirectory.Value)));
```
This does not mean data will be copied to memory. SO I am confused.

afchung replied:
Why doesn't it?

dkm2110 replied:
As per discussion offline, I think confusion was the comment to the function. Consider changing it to: 
"Deserialize and cache the materialized data in memory from the level the data is currently at ."

afchung replied:
Done.

dkm2110 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+    /// A class that manages the caching of data at different layers of storage medium.
+    /// </summary>
+    [Unstable("0.15", "API contract may change.")]
+    public sealed class DataCache<T>
```
Move private members after public.

afchung replied:
Done

dkm2110 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs:
```diff
@@ -0,0 +1,388 @@
+            Assert.Equal(CacheLevel.NotLocal, cache.CacheLevel);
+            Assert.False(cache.DiskDirectory.IsPresent());
+
+            // Test remote to memory stream.
```
Test remote to materialized?

afchung replied:
Good catch.

dkm2110 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs:
```diff
@@ -0,0 +1,388 @@
+            Assert.Equal(CacheLevel.NotLocal, cache.CacheLevel);
+            Assert.False(cache.Materialized.IsPresent());
+
+            // Test remote to materialized.
```
Test remote to memory?

afchung replied:
Done.

dkm2110 commented:
@afchung Left few minor and one major comment.

dkm2110 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        /// <summary>
+        /// "Materializes" and deserializes the data. Does not cache.
```
Does not cache the materialized data.

afchung replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs:
```diff
@@ -123,23 +131,23 @@ private ISet<string> Download()
         /// </summary>
         /// <returns></returns>
-        public T GetPartitionHandle()
+        public IEnumerable<T> GetPartitionHandle()
```
I remember for some reason, we changed from returning IEnumerable<T> to T when I wrote the original code. @dkm2110 Do you remember what was that requirement from IMRU?

afchung replied:
My reasoning for using `IEnumerable` is simple - it gives more flexibility. One can always return an `IEnumerable` of one item. It also works well with the new `ISerializer`, which only returns an object of `T` from a single `Stream` on deserialization. This means that you can easily match a `Stream` to a single object of `T`.

jwang98052 replied:
I agree it is clear. That was the reason I used it at the first place. But Dhruv asked for the change because of IMRU usage as I remembered. Just need to clarify from him.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs:
```diff
@@ -123,23 +131,23 @@ private ISet<string> Download()
 
                     // For now, assume IFileDeSerializer is local.
-                    return _fileSerializer.Deserialize(_localFiles.Value);
+                    return new List<T> { _fileSerializer.Deserialize(_localFiles.Value) };
```
How do you guarantee the type returned from Deserialize can be converted to IList<T>

afchung replied:
I'm not guarantee-ing it, I'm just putting whatever is returned into a `List`.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+                case CacheLevel.InMemoryMaterialized:
+                    CleanUpMaterialized(false);
+                    break;
+            }
```
No default?

afchung replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+                    _diskDirectory = Optional<IDirectoryInfo>.Of(_inputDataMover.RemoteToDisk());
+                    break;
+                default:
+                    throw new SystemException(
```
use Exceptions.Throw()

afchung replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,434 @@
+                        new List<MemoryStream>(_inputDataMover.RemoteToMemory()));
+                    break;
+                default:
+                    throw new SystemException(
```
use Exceptions.Throw()

afchung replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs:
```diff
@@ -0,0 +1,388 @@
+
+            public readonly IDirectoryInfo DirectoryInfo = Substitute.For<IDirectoryInfo>();
+
+            public DataCache<string> GetDataCache()
```
internal?

afchung replied:
`private` overrides `internal`. This is also a test class.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileStreamInputDataMover.cs:
```diff
@@ -0,0 +1,70 @@
+    /// of type T.
+    /// </summary>
+    [Unstable("0.15", "API contract may change.")]
+    public abstract class FileStreamInputDataMover<T> : SerializerBasedInputDataMover<T>
```
Is there any subclass of it that can be used in testing? Your tests use mock, Is this class got tested?

afchung replied:
This will be tested with [REEF-1397](https://issues.apache.org/jira/browse/REEF-1397), but this is mainly just a convenience abstract class. I didn't want to make the PR too big (and it's already 1200+ lines).

jwang98052 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs:
```diff
@@ -0,0 +1,388 @@
+        public void TestCleanCacheClearsAllUpperCacheLevels()
+        {
+            var cache = _context.GetDataCache();
+            SubstituteForBasicDataMover();
```
You might want to have a method like GetDataCachewithBasicDataMover() to wrap those two lines as they are used in multiple places.

afchung replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs:
```diff
@@ -0,0 +1,388 @@
+            SubstituteForBasicDataMover();
+            Assert.Equal(CacheLevel.Disk, cache.Cache(CacheLevel.Disk, false));
+            Assert.Equal(CacheLevel.InMemoryAsStream, cache.Cache(CacheLevel.InMemoryAsStream, false));
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryMaterialized, false));
```
The first three validation can be done in one test only. Rest of the test don't need to repeat those Assert but focus on what you want to test in that case.

afchung replied:
Given that this is an object that keeps state, I opt to do it for every test and every state because both the values passed to the object's methods and the order of method invocation/order of state change matters. Every test differs slightly in one (or more) of the above.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs:
```diff
@@ -173,7 +174,7 @@ public void TestWithByteDeserializer()
                         .GetInstance<IInputPartition<IEnumerable<byte>>>();
```
In this test, the T passed to IInputPartition is IEnumerable<byte> rather than byte. That would end up GetPartitionHandle returns  IEnumerable<IEnumerable<byte>>. Not sure how those tests are still working.

jwang98052 replied:
This pattern is used everywhere in the tests. If we decide to use T to represent a raw data type, like string or byte, then all those places in the tests need to be updated.

jwang98052 replied:
We also have tests in HadoopFileInputPartitionTest for Yarn testing. Those tests also need to be updated and tested on Yarn. They are exe, so can simply run on Yarn with different argument  for Run.

afchung replied:
I didn't want to break backward compatibility, so I opted to keep `T` as `IEnumerable`. We will be writing a new version of this class, and the test will be revised.

afchung replied:
Opened [REEF-1397](https://issues.apache.org/jira/browse/REEF-1397).

jwang98052 commented:
I completed a pass.

afchung commented:
@dkm2110 @jwang98052 I've addressed your comments, please have another look. Thanks!

markusweimer commented:
I'd like to have a look, but won't get to it before tonight.

markusweimer commented on lang/cs/Org.Apache.REEF.IO/CacheLevel.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <summary>
+    /// The level at which data is cached.
+    /// </summary>
+    public enum CacheLevel
```
Can you please add some `0`s to the constants in here? That way, we can slide `SSD` in between memory and disk at some point.

afchung replied:
Done.

markusweimer commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,479 @@
+        /// </summary>
+        public void Clear()
+        {
+            while (CleanCacheAtLevel(CacheLevel) != CacheLevel.NotLocal)
```
Isn't this a runaway loop?

afchung replied:
Could you clarify? This is always going to go to `CacheLevel.NotLocal`.

markusweimer commented on lang/cs/Org.Apache.REEF.IO/DataCache.cs:
```diff
@@ -0,0 +1,479 @@
+        /// </summary>
+        /// <param name="cacheToLevel">Level to cache to.</param>
+        /// <param name="shouldCleanHigherLevelCache">Whether or not to clean the higher level caches.</param>
+        /// <returns>The cached level.</returns>
```
How does this differ from `cacheToLevel`?

afchung replied:
It will return the lowest cache level if `_lowestCacheLevel` is already lower than or equal to than `cacheToLevel`.

markusweimer commented:
Done with a pass. Overall, LGTM from a code quality perspective. On the design, I have one concern: This enumerates all the cache levels explicitly in `public` APIs. This is fine as long as we don't see ourselves supporting more (or less) caching levels on a particular system. Are we sure of that?

An alternative would be to make a caching tier an interface that can move data into itself from another caching tier. The cache manager than would get a set of these, including a weight which indicates which cache is faster than which other cache. 

Am I making sense?

dkm2110 commented:
@markusweimer Just to make sure I understand: each caching option in the enum becomes an implementaiton of interface. Makes sense over long run. An alternative is to merge this one and then create another work item. I leave it to @afchung to decide.

markusweimer commented:
> each caching option in the enum becomes an implementaiton of interface.

Yes, that is the idea. It makes the memory hierarchy itself extensible. Not sure whether that is overkill, and I am not saying we should do this instead of this PR. It just occurred to me as an alternative design while doing the code review.

dkm2110 commented:
Got it... @afchung once you address 1-2 minor comments suggested by markus, I will merge.

dkm2110 commented:
@markusweimer as an examples of memory extensibility, GPU memory is example of another cache we might want to extend to later on :)

afchung commented:
@markusweimer @dkm2110 Thanks for the comments!

So, are we suggesting to merge this first? I am a bit worried that if we go with this current design it will just stick. If we don't envision supporting *a lot* more caching levels, I think this would be okay, but otherwise I would like to re-submit a PR.

I'm also thinking of changing the `interface` `IDataMover` to an `abstract` class with a base implementation that all throw `NotImplementedException` for every method, such that it will not break backward compatibility should we decide to extend it.

What do you think?

markusweimer commented:
> So, are we suggesting to merge this first?

I am fine either way. But with stuff like non-volatile memory complicating matters on the horizon, I'd be in favor to also work on the extensible API. 

Regarding the `abstract` class: I'm not in favor of these. They create maintenance issues down the line. We could always evolve the API via extension methods instead.

afchung commented:
@markusweimer extension methods are not overrideable and do not provide strong hints for the user to override them. How would an `abstract` class that only throws `NotImplementedException` create maintenance issues down the line?

afchung commented:
@markusweimer I've addressed your comments, please have another look. Would also be great if you could provide feedback as to why we shouldn't use `abstract` classes in this simple case to improve extensibility. Thanks!

markusweimer commented:
Abstract classes create a fragile base class problem. Pure abstract classes by definition don't, though.

dkm2110 commented:
@afchung @markusweimer  So, is it ready to merge? I will do it then.

jwang98052 commented:
LGTM. The only thing is public GetPartitionHandle() returns IEnumerable<T>. @dkm2110 do you have any comment on it? I may missed it in above discussions. I remember I returned IEnumerable<T> at the first time but changed it to return T as requested by you. Was there any special reason? 
To me return IEnumerable<T> is OK. But the follow up Jira must ensure the test cases are updated as all the test cases pass IEnumerable<int> or IEnumerable<byte>. Without the change, it will end up IEnumerable<IEnumerable<T>>.

markusweimer commented:
Good catch, @jwang98052! What is the reason to revert to `IEnumarable` here?

dkm2110 commented:
@jwang98052 @markusweimer At that point my thinking was that Generic type T itself can be IEnumerable<T> if needed. I had raised this in the earlier PR related to this JIRA to @afchung . I am pasting his response below from earlier PR:

"We need this because we would ideally like to be able to map a Stream to an object with the new ISerializer interface. The ISerializer does not have a natural way to map multiple files to a single T. This allows more flexibility - the user can always merge streams and return a single T if that is what they would like in their implementations."

markusweimer commented:
This is interesting. We have had two different interpretations of what the type `T` of a partition is:

  * It is a pointer to the real data. This is what we've done so far. `T` could be `IEnumarable<T2>` in that case.
  * It is the record type. This seems to be the mode we are going after now.

Of course, algos mostly care about their record type. However, I think exposing it like this here is problematic. If we expose it here, we end up prescribing many potentially differentiating approaches to this: How to manage memory? How to serialize data? Is it a stream or an `IEnumerable`? 

Hence, I'd favor a layered design as we had so far. Or does that close the door on other optimizations?

afchung commented:
I see, thanks for the explanation. I think either way is fine and won't necessarily preclude any optimizations, as an `IEnumerable<T>` can easily consist only of a single object and a `T` can also be an `IEnumerable` of some type of object. `T` here in this PR doesn't necessarily need to be a record type, it can also be a some pointer to a record type where you pass it through another transformation function to obtain the actual record type. @dkm2110 what was the original intention of the design?

For the `ISerializer` interface provided in this PR, one can implement a `Stream` that multiplexes between multiple `Stream` s. Either way, the muxing logic has to be implemented by the user. Personally I find it a bit unnatural for a Serializer to deserialize a collection of `Stream` s to a single object.

@markusweimer @dkm2110 @jwang98052 On another note, I've been working on another more extensible `DataCache<T>` (which uses some Reflection techniques that I don't really like...), but we'll still need to answer the question of what `T` is intended to be first.

I'll provide an update with a link to a branch with the new design in the JIRA once I am done. Thanks!

dkm2110 commented:
@afchung So as a ML algorithm developer, I  access data like IEnumerable<Record> . There will always be some Data class that will encapsulate this Enumerable and will have other functions  - NumberOfPoints, DataTimesVector etc. are some examples. So in my mind when I do GetPartitionHandle() in my algorithms for example mapper in IMRU, I will get this Data class than enumerable only. And as you said, nothing stops us from T being IEnumerable<Record> anyways.

dkm2110 commented:
Regarding ISerializer, as a user I do not care if data is in multiple streams or files. What I see at the end is a single data class that will read from these multiple streams using whatever logic.

dkm2110 commented:
Also another interesting case that I encountered just now is to extend cache layer API to any variable and not necessarily input only. For example, consider simple K-Means. For big data if number of centers and dimensionality both are big, I would like to store the centers as chunks on disk and then when needed load and use them chunk by chunk.

afchung commented:
@markusweimer @dkm2110 @jwang98052 thanks the review and for all the helpful comments. Let me close this PR first such that it doesn't accidentally get merged. We should converge on design on the JIRA item first. I'll post up a link to my new design shortly. Thanks!

dkm2110 commented:
@afchung Thanks...However, this whole process has been very helpful since this caching aspect is extremely important.
